packages/dds/shared-object-base/src/sharedObject.ts

```diff
@@ -519,3 +447,118 @@ export abstract class SharedObject<TEvent extends ISharedObjectEvents = ISharedO
+        /**
+         * During summarize, the SummarySerializer keeps track of IFluidHandles that are serialized. These handles
+         * represent references to other Fluid objects and are used for garbage collection.
+         *
+         * This is fine for now. However, if we implement delay loading in DDss, they may load and de-serialize content
+         * in summarize. When that happens, they may incorrectly hit this assert and we will have to change this.
+         */
+        assert(!this._isSummarizing,
+            0x075 /* "SummarySerializer should be used for serializing data during summary." */);
+        return super.serializer;
+    }
+
+    /**
+     * @param id - The id of the shared object
+     * @param runtime - The IFluidDataStoreRuntime which contains the shared object
+     * @param attributes - Attributes of the shared object
+    */
+    constructor(
+        public id: string,
+        protected runtime: IFluidDataStoreRuntime,
+        public readonly attributes: IChannelAttributes)
+    {
+        super(id, runtime, attributes);
+    }
+
+    /**
+     * {@inheritDoc (ISharedObject:interface).summarize}
+     */
+    public summarize(fullTree: boolean = false, trackState: boolean = false): ISummaryTreeWithStats {
+        // Set _isSummarizing to true. This flag is used to ensure that we only use SummarySerializer (created below)
+        // to serialize handles in this object's data. The routes of these serialized handles are outbound routes
+        // to other Fluid objects.
+        assert(!this._isSummarizing, 0x076 /* "Possible re-entrancy! Summary should not already be in progress." */);
+        this._isSummarizing = true;
+
+        let summaryTree: ISummaryTreeWithStats;
+        try {
+            const serializer = new SummarySerializer(
+                this.runtime.channelsRoutingContext,
+                (handle: IFluidHandle) => this.handleDecoded(handle),
+            );
+            const snapshot: ITree = this.snapshotCore(serializer);
+            summaryTree = convertToSummaryTreeWithStats(snapshot, fullTree);
+            assert(this._isSummarizing, 0x077 /* "Possible re-entrancy! Summary should have been in progress." */);
+        } finally {
+            this._isSummarizing = false;
+        }
+        return summaryTree;
+    }
+
+    public async summarizeAsync(fullTree: boolean = false, trackState: boolean = false):
+        Promise<ISummaryTreeWithStats> {
+        return this.summarize(fullTree, trackState);
+    }
+
+    /**
+     * {@inheritDoc (ISharedObject:interface).getGCData}
+     */
+    public getGCData(fullGC: boolean = false): IGarbageCollectionData {
+        // Set _isSummarizing to true. This flag is used to ensure that we only use SummarySerializer (created in
+        // getGCDataCore) to serialize handles in this object's data.
+        assert(!this._isSummarizing, 0x078 /* "Possible re-entrancy! Summary should not already be in progress." */);
```

Comment:
I think this code should move into getGCDataCore() because getGCDataCore() is the one who sets up and uses custom serializer

Reply:
I agree the serializer functionality should be consolidated in one place. Considering that getGCDataCore is meant to be overridden for complex DDSs, I think it would make more sense to put them both in getGCData and pass the serializer to getGCDataCore. Although it's possible that an implementation of getGCDataCore wouldn't use a serializer at all.